File: kie-wb-common-services/kie-wb-common-verifier/kie-wb-common-verifier-api/src/test/java/org/kie/workbench/common/services/verifier/api/client/MarshallingTest.java
/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kie.workbench.common.services.verifier.api.client;

import java.util.Collections;

import org.drools.verifier.api.reporting.CheckType;
import org.drools.verifier.api.reporting.ImpossibleMatchIssue;
import org.drools.verifier.api.reporting.Issue;
import org.drools.verifier.api.reporting.MultipleValuesForOneActionIssue;
import org.drools.verifier.api.reporting.RedundantConditionsIssue;
import org.drools.verifier.api.reporting.Severity;
import org.drools.verifier.api.reporting.SingleHitLostIssue;
import org.drools.verifier.api.reporting.ValueForActionIsSetTwiceIssue;
import org.drools.verifier.api.reporting.ValueForFactFieldIsSetTwiceIssue;
import org.jboss.errai.marshalling.server.MappingContextSingleton;
import org.jboss.errai.marshalling.server.ServerMarshalling;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

public class MarshallingTest {

    @Before
    public void prepare() throws Exception {
        MappingContextSingleton.get();
    }

    @Test
    public void testImpossibleMatchIssue() throws Exception {
        check(new ImpossibleMatchIssue(Severity.WARNING,
                                       CheckType.IMPOSSIBLE_MATCH,
                                       "1",
                                       "String",
                                       "name",
                                       "1",
                                       "2",
                                       Collections.singleton(1)),
              new ImpossibleMatchIssue(Severity.ERROR,
                                       CheckType.IMPOSSIBLE_MATCH,
                                       "1",
                                       "Person",
                                       "name",
                                       "1",
                                       "2",
                                       Collections.singleton(1)));
    }

    @Test
    public void testMultipleValuesForOneActionIssue() throws Exception {
        check(new MultipleValuesForOneActionIssue(Severity.ERROR,
                                                  CheckType.IMPOSSIBLE_MATCH,
                                                  "1",
                                                  "2",
                                                  Collections.emptySet()),
              new MultipleValuesForOneActionIssue(Severity.WARNING,
                                                  CheckType.IMPOSSIBLE_MATCH,
                                                  "1",
                                                  "2",
                                                  Collections.emptySet()));
    }

    @Test
    public void testRedundantConditionsIssue() throws Exception {
        check(new RedundantConditionsIssue(Severity.ERROR,
                                           CheckType.IMPOSSIBLE_MATCH,
                                           "Person",
                                           "name",
                                           "1",
                                           "2",
                                           Collections.emptySet()),
              new RedundantConditionsIssue(Severity.WARNING,
                                           CheckType.IMPOSSIBLE_MATCH,
                                           "Person",
                                           "name",
                                           "1",
                                           "2",
                                           Collections.emptySet()));
    }

    @Test
    public void testSingleHitLostIssue() throws Exception {
        check(new SingleHitLostIssue(Severity.ERROR,
                                     CheckType.IMPOSSIBLE_MATCH,
                                     "1",
                                     "2"),
              new SingleHitLostIssue(Severity.ERROR,
                                     CheckType.IMPOSSIBLE_MATCH,
                                     "2",
                                     "3"));
    }

    @Test
    public void testValueForActionIsSetTwiceIssue() throws Exception {
        check(new ValueForActionIsSetTwiceIssue(Severity.ERROR,
                                                CheckType.IMPOSSIBLE_MATCH,
                                                "1",
                                                "2",
                                                Collections.emptySet()),
              new ValueForActionIsSetTwiceIssue(Severity.WARNING,
                                                CheckType.IMPOSSIBLE_MATCH,
                                                "1",
                                                "2",
                                                Collections.emptySet()));
    }

    @Test
    public void testValueForFactFieldIsSetTwiceIssue() throws Exception {
        check(new ValueForFactFieldIsSetTwiceIssue(Severity.ERROR,
                                                   CheckType.IMPOSSIBLE_MATCH,
                                                   "boundName",
                                                   "name",
                                                   "1",
                                                   "2",
                                                   Collections.emptySet()),
              new ValueForFactFieldIsSetTwiceIssue(Severity.WARNING,
                                                   CheckType.IMPOSSIBLE_MATCH,
                                                   "boundName",
                                                   "name",
                                                   "1",
                                                   "2",
                                                   Collections.emptySet()));
    }

    private void check(final Issue original,
                       final Issue secondOne) {

        final String json = ServerMarshalling.toJSON(original);

        final Issue newVersion = (Issue) ServerMarshalling.fromJSON(json);

        assertEquals(original.hashCode(),
                     newVersion.hashCode());
        assertNotEquals(secondOne.hashCode(),
                        newVersion.hashCode());
    }
}

File: appformer-js/src/marshalling/__tests__/Marshalling.test.ts
/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MarshallerProvider } from "../MarshallerProvider";
import * as Marshalling from "../Marshalling";
import { unmarshall } from "../Marshalling";
import { JavaInteger } from "../../java-wrappers";
import { NumValBasedErraiObject } from "../model/NumValBasedErraiObject";
import { JavaType } from "../../java-wrappers/JavaType";

const originalGetForObject = MarshallerProvider.getForObject;
const originalGetForFqcn = MarshallerProvider.getForFqcn;

afterEach(() => {
  // reset static function, otherwise other tests will be invoking the function set by this test
  MarshallerProvider.getForFqcn = originalGetForFqcn;
  MarshallerProvider.getForObject = originalGetForObject;
});

describe("marshall", () => {
  test("with regular input, should return an errai object json-string version of it", () => {
    const input = new JavaInteger("1");
    const inputErraiObject = new NumValBasedErraiObject(JavaType.INTEGER, 1).asErraiObject();

    const expectedJson = JSON.stringify(inputErraiObject);

    // skip actual marshaller implementation
    const mockedMarshaller = { marshall: jest.fn(() => inputErraiObject), unmarshall: jest.fn() };
    MarshallerProvider.getForObject = jest.fn(() => mockedMarshaller);

    // ==
    // ====== test
    const output = Marshalling.marshall(input);

    // == assertion
    expect(expectedJson).toEqual(output);
  });

  test("with null input, should return null", () => {
    const input = null as any;

    const output = Marshalling.marshall(input);

    expect(output).toBeNull();
  });

  test("with undefined input, should serialize marshaller output", () => {
    const input = undefined as any;

    const output = Marshalling.marshall(input);

    expect(output).toBeNull();
  });
});

describe("unmarshall", () => {
  test("with null input, should return undefined", () => {
    const input = null;

    const output = unmarshall(input as any, new Map());

    expect(output).toBeUndefined();
  });

  test("with undefined input, should return undefined", () => {
    const input = undefined;

    const output = unmarshall(input as any, new Map());

    expect(output).toBeUndefined();
  });

  test("with invalid json, should throw error", () => {
    const input = "{ broken json }";

    expect(() => unmarshall(input, new Map())).toThrowError();
  });

  test("with regular input, should return marshallers' specific content", () => {
    const inputJson = new NumValBasedErraiObject(JavaType.INTEGER, 1).asErraiObject();

    const expectedOutput = new JavaInteger("1");

    // skip actual marshaller implementation
    const mockedMarshaller = { unmarshall: jest.fn(() => expectedOutput), marshall: jest.fn() };
    MarshallerProvider.getForFqcn = jest.fn(() => mockedMarshaller);

    // ==
    // ====== test
    const output = Marshalling.unmarshall(JSON.stringify(inputJson), new Map());

    // == assertion
    expect(expectedOutput).toEqual(output);
  });
});
